src/rust/engine/src/scheduler.rs

```diff
@@ -474,6 +476,19 @@ impl Scheduler {
     session.maybe_display_initialize(&self.core.executor, &sender);
     self.execute_helper(request, session, sender);
     let result = loop {
+      match externs::call(&python_signal, &[]) {
+        Ok(value) => {
+          if externs::is_truthy(&*value) {
+            return Err(ExecutionTermination::KeyboardInterrupt);
+          }
+        }
+        Err(e) => {
```

Comment:
If this function is running in the main thread (it won't always be), it's possible that the error type received here will actually be [KeyboardInterrupt](http://dgrunwald.github.io/rust-cpython/doc/cpython/exc/struct.KeyboardInterrupt.html): in which case you'd want to return `ExecutionTermination::KeyboardInterrupt`.